### PR TITLE
2.0

### LINKED ISSUE #2: Dynamic supervisor children
These are currently handwavely-typed. 

Quoth rob

> SimpleOneForOne means we lose the type
We could just use OneForOne instead and pass in the whole spec and then we'd be typesafe
and we just get rid of childStartTemplate so you *always* pass in the startLink *and* the args and get the safety
```

### PR DESCRIPTION
**Pull request by robashton**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,34 +1,19 @@
...
-		--docgen Pinto.Sup:docs/Pinto.Sup.md \
-		--docgen Pinto.App:docs/Pinto.App.md \
-		--docgen Pinto.Gen:docs/Pinto.Gen.md \
-		--docgen Pinto.Timer:docs/Pinto.Timer.md \
-		--docgen Pinto.Types:docs/Pinto.Types.md 
-	touch docs
-
-.psc-package: psc-package.json
-	psc-package install
-	touch .psc-package
```
Comment (robashton) [inline Makefile:30]: I put a lot of effort into writing the doc tags or whatever they're called for Pinto (editor support, docs into pursuit and everything). 

I would like to request very loudly that you re-instate this work when you're done.

File: test/Main.purs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    let
+      worked = case slRes of
+        Right pid -> true
+        Left reason -> false
+    assert worked
+    pure unit
+
+    where
+      init = do
+        pure $ Right $ InitOk $ TestState 0
```
Comment (robashton) [inline test/Main.purs]: Right *and* InitOk looks a little odd...

File: src/Pinto/GenServer.purs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+  | InitOkHibernate state
+
+data ServerNotRunning
+  = InitStop Foreign
+  | InitIgnore
+
+type InitResult cont state = Either ServerNotRunning (ServerRunning cont state )
+
+-- Can't do a functor instance over a type synonym, so just have a function instead
+mapInitResult :: forall state state' cont. (state -> state') -> InitResult cont state -> InitResult cont state'
```
Comment (robashton) [inline src/Pinto/GenServer.purs]: I can understand wanting to capture "success" vs "failure" in the init results, but the introduction of Either on top of an ADT doesn't really add anything unless you're using 'either' functionality somewhere *on* that result - @nwolverson ?

File: Makefile (hunk 1 above)
Comment (robashton) [reply Makefile:30]: Also https://github.com/robashton/purerl-cookbook :)

Comment (nwolverson) [reply Makefile:30]: Generally I wouldn't generate docs into `docs/` unless you're using that for something else, we should ensure that is updated on pursuit (fork) - generally all PureScript libraries used to have generated docs in the repo and moved to rm those and only host on pursuit - with the `purs docs` format outputs just for local/offline doc generation

File: src/Pinto/GenServer.purs (hunk 3 above)
Comment (nwolverson) [reply src/Pinto/GenServer.purs]: Yep that looks odd - if the ADT is representing init results it seems like it's just one thing, and stick a function `:: InitResult -> Either _ state` if required on top of that

File: Makefile (hunk 1 above)
Comment (robashton) [reply Makefile:30]: @nwolverson Oh sure, I was only doing that to test that I'd done it proper and that it was actually committed for some reason was entirely by the by

File: test/GenServer.erl [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+%%% ----------------------------------------------------------------------------
+%%% FFI API
+%%% ----------------------------------------------------------------------------
+startGprocFFI() ->
+  fun() ->
+      %% TODO - not this, anything but this
+      code:add_patha("/Users/adrianroe/dev/rtsv2/_build/default/lib/gproc/ebin/"),
+      application:ensure_started(gproc),
+      unit
+  end.
```
Comment (robashton) [inline test/GenServer.erl]: You can presumably now use the actual rebar3 eunit test runner with before/after actions to get this sort of thing working properly

**Comment by robashton** [comment]
I know the monitor implementation wasn't the greatest in the world, but I'm using that quite heavily in other projects - are you going to replace it?

**Comment by robashton** [comment]
Ditto Timer - that went through a lot of iterations and the solution ended up having to be a pragmatic one - that'll need to come back in some form also

**Comment by philipstears** [comment]
Yup, they are coming back, we just got rid of them temporarily whilst we were experimenting with replumbing various types.

File: test/DoorLock.purs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    handleEvent = unsafeCoerce unit
+
+    handleEnter Locked UnlockedClosed currentData = do
+      audit AuditDoorUnlocked
+      pure $ Right $ KeepStateAndData
+
+    handleEnter UnlockedOpen UnlockedClosed currentData = do
+      audit AuditDoorClosed
+      pure $ Right $ KeepStateAndData
```
Comment (robashton) [inline test/DoorLock.purs]: Again, this whole Right/Left thing around our ADTs doesn't look right

Comment (philipstears) [reply test/DoorLock.purs]: This is an older commit, the Right/Left distinction here was just copy/pasta, it's not there in the current version, except in the init function.

As to whether it should be there in the init function and in the StartLinkResult, I ended up bowing out of that debate, others had fairly strong opinions there and I was pretty much on the fence about the whole thing.

Comment (robashton) [reply test/DoorLock.purs]: :ok:

Comment (robashton) [reply test/DoorLock.purs]: er, :ok_hand:  even

File: test/DoorLock.purs [hunk 6]
```diff
@@ -1,38 +1,119 @@
...
-  | UnlockedClosed
-  | UnlockedOpen
+  = Locked { failedAttempts :: Int }
+  | UnlockedClosed { failedAttemptsBeforeUnlock :: Int }
+  | UnlockedOpen { failedAttemptsBeforeUnlock :: Int }
+
+instance stateHasStateId :: HasStateId StateId State where
+  getStateId (Locked _) = StateIdLocked
+  getStateId (UnlockedClosed _) = StateIdUnlockedClosed
+  getStateId (UnlockedOpen _) = StateIdUnlockedOpen
```
Comment (robashton) [inline test/DoorLock.purs:89]: I think I like this

**Pull request closed without merging**